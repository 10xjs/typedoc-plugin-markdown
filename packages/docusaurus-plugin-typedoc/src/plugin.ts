import * as path from 'path';

import { LoadContext } from '@docusaurus/types';
import { Application } from 'typedoc';

import { DocsaurusFrontMatterComponent } from './components/front-matter.component';
import { writeSidebar } from './sidebar';
import DocusaurusTheme from './theme/theme';
import { PluginOptions } from './types';

const DEFAULT_PLUGIN_OPTIONS: PluginOptions = {
  inputFiles: ['../src/'],
  docsRoot: 'docs',
  out: 'api',
  sidebar: {
    fullNames: false,
    sidebarFile: 'typedoc-sidebar.js',
    globalsLabel: 'Globals',
    readmeLabel: 'README',
  },
  plugin: ['typedoc-plugin-markdown'],
};

let app: Application;

export default function pluginDocusaurus(
  context: LoadContext,
  opts: Partial<PluginOptions>,
) {
  const { siteDir } = context;

  /**
   * Configure options
   */
  const options = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...opts,
    // deep merge plugin options
    ...(opts.plugin && {
      plugin: [
        ...['typedoc-plugin-markdown'],
        ...opts.plugin.filter((name) => name !== 'typedoc-plugin-markdown'),
      ],
    }),
    // deep merge sidebar
    ...(opts.sidebar && {
      sidebar: {
        ...DEFAULT_PLUGIN_OPTIONS.sidebar,
        ...opts.sidebar,
      },
    }),
  };

  /**
   * Initialize and build app
   * - currently we can't re-compile with devServer due to infinate render loop
   * - @TODO look if configureWebpack lifecycle can be updated to handle more gracefully
   */
  if (!app) {
    app = new Application();

    // configure typedoc options (remove docusaurus props and pass everything else to renderer)
    const typedocOptions = Object.keys(options).reduce((option, key) => {
      if (!['id', 'inputFiles', 'sidebar', 'out', 'docsRoot'].includes(key)) {
        option[key] = options[key];
      }
      return option;
    }, {});

    // bootstrap typedoc app (pass in docusaurus theme file)
    app.bootstrap({
      ...typedocOptions,
      theme: path.resolve(__dirname, 'theme'),
    });

    // add frontmatter component
    app.renderer.addComponent(
      'docusaurus-frontmatter',
      new DocsaurusFrontMatterComponent(
        app.renderer,
        options.out,
        options.sidebar,
      ),
    );

    // return the generated reflections
    const project = app.convert(app.expandInputFiles(options.inputFiles));

    // if project is undefined typedoc has a problem - error logging will be supplied by typedoc.
    if (!project) {
      return;
    }

    // construct outputDirectory path
    const outputDirectory = path.resolve(
      siteDir,
      options.docsRoot,
      options.out,
    );

    // generate the static docs
    app.generateDocs(project, outputDirectory);

    // write the sidebar (if applicable)
    if (options.sidebar) {
      const theme = app.renderer.getComponent('theme') as DocusaurusTheme;
      writeSidebar(
        options.disableOutputCheck || theme.isOutputDirectory(outputDirectory),
        siteDir,
        options.out,
        options.sidebar,
        theme.getNavigation(project),
      );
    }
  }

  // we need to generate the sidebar before any available lifecycle apis
  return {
    name: 'docusaurus-plugin-typedoc',
  };
}
