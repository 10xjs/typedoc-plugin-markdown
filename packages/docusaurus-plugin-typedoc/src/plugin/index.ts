import * as path from 'path';

import { LoadContext, Plugin } from '@docusaurus/types';
import * as fs from 'fs-extra';
import { Application } from 'typedoc';
import { ModuleKind, ScriptTarget } from 'typescript';

export default function pluginDocusaurus(
  context: LoadContext,
  options: Partial<any>,
): Plugin<any> {
  const { siteDir } = context;

  const inputFiles = options.inputFiles;
  const docsRoot = path.resolve(siteDir, 'docs');
  const outFolder = options.out ? options.out : undefined;
  const out = docsRoot + (options.out ? '/' + options.out : '');
  const skipSidebar = options.skipSidebar || false;

  delete options.skipSidebar;
  delete options.inputFiles;
  delete options.out;

  return {
    name: 'docusaurus-plugin-typedoc',

    async loadContent() {
      try {
        const app = new Application();
        app.bootstrap({
          module: ModuleKind.CommonJS,
          target: ScriptTarget.ES5,
          disableOutputCheck: true,
          readme: 'none',
          plugin: ['typedoc-plugin-markdown'],
          theme: path.resolve(__dirname, '..', 'theme'),
          ...options,
        });
        const project = app.convert(app.expandInputFiles(inputFiles));
        app.generateDocs(project, out);
        if (!skipSidebar) {
          const sidebarPath = path.resolve(siteDir, 'sidebars.js');
          const navigation = app.renderer.theme.getNavigation(project);
          const sidebarContent = getNavObject(outFolder, navigation);
          writeSideBar(sidebarContent, sidebarPath);
        }
      } catch (e) {
        return;
      }
    },
  };
}

function getNavObject(outFolder, navigation) {
  const navObject = {};
  let url = '';
  let navKey = '';
  navigation.children.forEach((rootNavigation) => {
    rootNavigation.children.forEach((item) => {
      url = item.url.replace('.md', '');
      navKey = url.substr(0, url.indexOf('/'));
      if (navKey !== undefined && navKey.length) {
        navKey = navKey[0].toUpperCase() + navKey.slice(1);
      }
      if (navObject[navKey] === undefined) {
        navObject[navKey] = [];
      }
      const pageUrl = outFolder ? outFolder + '/' + url : url;
      if (!navObject[navKey].includes(pageUrl)) {
        navObject[navKey].push(pageUrl);
      }
    });
  });
  return { typedoc: navObject };
}

function writeSideBar(navObject: any, sidebarPath: string) {
  let jsonContent: any;
  if (!fs.existsSync(sidebarPath)) {
    jsonContent = JSON.parse('{}');
  } else {
    jsonContent = require(sidebarPath);
  }

  jsonContent = Object.assign({}, jsonContent, navObject);
  try {
    fs.outputFileSync(
      sidebarPath,
      'module.exports = ' + JSON.stringify(jsonContent, null, 2) + ';',
    );
    console.log(
      `[docusaurus-plugin-typedoc] sidebar updated at ${sidebarPath}`,
    );
  } catch (e) {
    console.log(
      `[docusaurus-plugin-typedoc] failed to update sidebar at ${sidebarPath}`,
    );
  }
}
