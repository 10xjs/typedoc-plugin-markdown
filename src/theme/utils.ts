import * as fs from 'fs';
import * as Handlebars from 'handlebars';
import * as path from 'path';
import { DeclarationReflection, Reflection } from 'typedoc/dist/lib/models/reflections/index';

export function compileTemplate(template: string, data: {}) {
  const source = templateToString(template);
  const md: HandlebarsTemplateDelegate = Handlebars.compile(source);
  return new Handlebars.SafeString(md(data));
}

export function safeString(data: string) {
  return new Handlebars.SafeString(data);
}

export function getAnchorRef(ref: string) {
    return ref.replace(/_|\/|\.| /g, '-').replace(/"/g, '').replace(/ /g, '-').toLowerCase();
  }

function templateToString(template: string) {
  return fs.readFileSync(path.join(__dirname, `/${template}`)).toString();
}
