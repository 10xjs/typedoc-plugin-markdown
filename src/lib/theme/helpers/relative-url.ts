import * as path from 'path';
import { MarkdownPlugin } from '../../plugin';

export function relativeUrl(absolute: string) {
  const urlPrefix: RegExp = /^(http|ftp)s?:\/\//;
  if (!MarkdownPlugin.location || urlPrefix.test(absolute)) {
    return absolute;
  } else {
    const relative = path.relative(path.dirname(MarkdownPlugin.location), path.dirname(absolute));
    return path.join(relative, path.basename(absolute)).replace(/\\/g, '/');
  }
}
