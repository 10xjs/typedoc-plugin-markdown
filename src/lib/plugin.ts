import * as path from 'path';
import { Application, DeclarationReflection, ProjectReflection } from 'typedoc';
import { Converter } from 'typedoc/dist/lib/converter';
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Reflection } from 'typedoc/dist/lib/models/reflections/abstract';
import { PageEvent, RendererEvent } from 'typedoc/dist/lib/output/events';
import { Renderer } from 'typedoc/dist/lib/output/renderer';
import { Options } from 'typedoc/dist/lib/utils/options';

import { MarkdownTheme } from './theme/theme';
import { BitbucketTheme } from './theme/theme.bitbucket';
import { DocusaurusTheme } from './theme/theme.docusaurus';
import { GitbookTheme } from './theme/theme.gitbook';

@Component({ name: 'markdown' })
export class MarkdownPlugin extends ConverterComponent {
  static theme: MarkdownTheme | DocusaurusTheme;
  static disableOutputCheck: boolean;
  static application: Application;
  static reflection: Reflection;
  static location: string;
  static page: PageEvent;
  static settings: {
    platform?: 'gitbook' | 'ducusaurus' | 'bitbucket';
    hideSources?: boolean;
    readme?: string;
    includes?: string;
    media?: string;
    mode?: string;
    mdHideSources?: boolean;
    mdSourceRepo?: string;
  };
  static project: ProjectReflection;

  initialize() {
    this.listenTo(this.owner, {
      [Converter.EVENT_RESOLVE_BEGIN]: this.onBegin,
    });
    this.listenTo(this.application.renderer, {
      [PageEvent.BEGIN]: this.onPageBegin,
      [PageEvent.END]: this.onPageEnd,
      [RendererEvent.BEGIN]: this.onRenderBegin,
    });
  }

  onBegin() {
    MarkdownPlugin.application = this.application;
    MarkdownPlugin.settings = this.application.options.getRawValues();
    MarkdownPlugin.setTheme(this.application.renderer, this.application.options);
  }

  onRenderBegin(renderer: RendererEvent) {
    if (MarkdownPlugin.theme) {
      MarkdownPlugin.project = renderer.project;
    }
  }

  onPageBegin(page: PageEvent) {
    if (MarkdownPlugin.theme) {
      MarkdownPlugin.reflection = page.model instanceof DeclarationReflection ? page.model : undefined;
      MarkdownPlugin.location = page.model.url ? page.model.url : '';
    }
  }

  onPageEnd(page: PageEvent) {
    if (MarkdownPlugin.theme) {
      page.contents = page.contents ? MarkdownPlugin.formatContents(page.contents) : '';
    }
  }

  static setTheme(renderer: Renderer, options: Options) {
    const themeName = options.getValue('theme');
    const themePath = path.join(__dirname, './theme/');
    const media = options.getValue('media');
    const platform = options.getValue('platform') || options.getValue('mdEngine');
    const theme = themeName === 'markdown' ? this.getTheme(platform, renderer, themePath, options) : null;
    if (theme) {
      if (media && (theme instanceof DocusaurusTheme || theme instanceof GitbookTheme)) {
        MarkdownPlugin.application.logger.warn(`[typedoc-markdown-plugin] media option is currently not supported in ${platform} theme`);
        options.setValue('media', null);
      }
      renderer.theme = renderer.addComponent('theme', theme);
      MarkdownPlugin.theme = theme;
    } else {
      MarkdownPlugin.application.logger.write('Markdown theme not set');
    }
  }

  static getTheme(platform: string, renderer: Renderer, themePath: string, options: any) {
    if (platform) {
      if (platform === 'gitbook') {
        return new GitbookTheme(renderer, themePath, options);
      }
      if (platform === 'docusaurus') {
        return new DocusaurusTheme(renderer, themePath, options);
      }
      if (platform === 'bitbucket') {
        return new BitbucketTheme(renderer, themePath, options);
      }
    }
    return new MarkdownTheme(renderer, themePath, options);
  }

  static formatContents(contents: string) {
    return contents
      .replace(/^ +/gm, '')
      .replace(/[\r\n]{3,}/g, '\n\n')
      .replace(/!spaces/g, '')
      .trim();
  }
}
