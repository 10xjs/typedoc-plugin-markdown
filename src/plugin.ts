/**
 * Plugin definition
 */
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';
import { Context } from 'typedoc/dist/lib/converter/context';
import { Converter } from 'typedoc/dist/lib/converter/converter';
import { Reflection } from 'typedoc/dist/lib/models/reflections/abstract';
import { Options, OptionsReadMode } from 'typedoc/dist/lib/utils/options';
import { MarkdownTheme } from './theme/theme';

import * as path from 'path';

@Component({ name: 'markdown' })
export class MarkdownPlugin extends ConverterComponent {

  private isSinglePage: boolean;

  public initialize() {

    const options: Options = this.application.options;
    options.read({}, OptionsReadMode.Prefetch);
    this.isSinglePage = (options.getValue('markdownSinglePage'));

    this.listenTo(this.owner, {
      [Converter.EVENT_RESOLVE_BEGIN]: this.onBegin,
    });
  }

  private getThemeDirectory() {
    return path.join(__dirname, '../dist/theme/');
  }

  private onBegin(context: Context, reflection: Reflection) {

    const renderer = this.application.renderer;
    const theme = this.application.options.getValue('theme');
    const themePath = this.getThemeDirectory();
    const options = { isSinglePage: this.isSinglePage };

    // apply the theme
    if (theme === 'markdown') {
      renderer.theme = renderer.addComponent('theme', new MarkdownTheme(this.application.renderer, themePath, options));
 } else {
      this.application.logger.log('To generate markdown please set option --theme markdown');
    }
  }

}
