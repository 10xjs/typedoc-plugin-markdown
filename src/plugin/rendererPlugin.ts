import * as FS from 'fs-extra';
import * as Handlebars from 'handlebars';
import * as Path from 'path';

import { Component, ContextAwareRendererComponent } from 'typedoc/dist/lib/output/components';
import { RendererEvent } from 'typedoc/dist/lib/output/events';

import { Options } from '../theme/options';

@Component({ name: 'markdown-renderer' })
export class MarkdownRendererPlugin extends ContextAwareRendererComponent {

  /**
   * Path to the output media directory.
   */
  private mediaDirectory: any;

  /**
   * The pattern used to find references in markdown.
   */
  private includePattern: RegExp = /\[\[include:([^\]]+?)\]\]/g;

  /**
   * The pattern used to find media links.
   */
  private mediaPattern: RegExp = /media:\/\/([^ "\)\]\}]+)/g;

  /**
   * Create a new MarkedPlugin instance.
   */
  public initialize() {
    super.initialize();

    const that = this;
    Handlebars.registerHelper('parseComments', function(arg: any) { return that.parseComments(arg.fn(this), this); });
    Handlebars.registerHelper('relativeURL', (url: string) => url ? this.getRelativeUrl(url) : url);

  }

  /**
   * Parse the given markdown string and return the resulting html.
   *
   * @param text  The markdown string that should be parsed.
   * @param context  The current handlebars context.
   * @returns The resulting html string.
   */
  public parseComments(text: string, context: any) {

    if (Options.includes) {
      text = text.replace(this.includePattern, (match: string, path: string) => {
        path = Path.join(Options.includes, path.trim());
        if (FS.existsSync(path) && FS.statSync(path).isFile()) {
          const contents = FS.readFileSync(path, 'utf-8');
          if (path.substr(-4).toLocaleLowerCase() === '.hbs') {
            const template = Handlebars.compile(contents);
            return template(context);
          } else {
            return contents;
          }
        } else {
          return '';
        }
      });
    }

    if (this.mediaDirectory) {
      text = text.replace(this.mediaPattern, (match: string, path: string) => {
        if (FS.existsSync(Path.join(this.mediaDirectory, path))) {
          return this.getRelativeUrl('media') + '/' + path;
        } else {
          return match;
        }
      });
    }

    return text;
  }

  protected onBeginRenderer(event: RendererEvent) {
    super.onBeginRenderer(event);

    if (Options.media) {
      const media = Path.resolve(Options.media);
      if (FS.existsSync(media) && FS.statSync(media).isDirectory()) {
        this.mediaDirectory = Path.join(event.outputDirectory, 'media');
        FS.copySync(media, this.mediaDirectory);
      } else {
        this.mediaDirectory = null;
      }
    }
  }

}
