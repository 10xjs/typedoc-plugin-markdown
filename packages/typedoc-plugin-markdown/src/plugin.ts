import * as path from 'path';

import { Renderer } from 'typedoc';
import { Converter } from 'typedoc/dist/lib/converter';
import { Component, ConverterComponent } from 'typedoc/dist/lib/converter/components';

@Component({ name: 'markdown' })
export class MarkdownPlugin extends ConverterComponent {
  initialize() {
    this.listenTo(this.owner, {
      [Converter.EVENT_BEGIN]: this.onBegin,
      [Converter.EVENT_RESOLVE_BEGIN]: this.onResolveBegin,
    });
  }

  /**
   * Overide the default assets for any custom themes to inherit
   */
  onBegin() {
    Renderer.getDefaultTheme = () => path.join(__dirname, 'resources');
  }

  /**
   * Read the theme option and load the paths of any recognised built in themes
   * Otherwise pass the path through to the Renderer
   */
  onResolveBegin() {
    const options = this.application.options;
    const theme = options.getValue('theme') as string;

    // messaging regarding refactored themes
    const messagePrefix = (theme: string) => {
      return `[typedoc-plugin-markdown] Please note the ${theme} theme is no longer supported in v3. `;
    };

    if (['docusaurus', 'docusaurus2'].includes(theme)) {
      this.application.logger.warn(messagePrefix('docusaurus') + 'Please use docusaurus-plugin-typedoc.');
    }
    if (theme === 'vuepress') {
      this.application.logger.warn(messagePrefix('vuepress') + 'Please use vuepress-plugin-typedoc.');
    }
    if (theme === 'bitbucket') {
      this.application.logger.warn(
        messagePrefix('bitbucket') + 'Please use --bitbucketCloudAnchors option to fix anchor links.',
      );
    }
    if (theme === 'gitbook') {
      this.application.logger.warn(messagePrefix('gitbook'));
    }

    const themes = ['default', 'markdown', 'docusaurus', 'docusaurus2', 'bitbucket', 'vuepress', 'gitbook'];

    // if the theme is 'default' or 'markdown' load the base markdown theme
    if (themes.includes(theme)) {
      options.setValue('theme', path.join(__dirname));
    }
  }
}
