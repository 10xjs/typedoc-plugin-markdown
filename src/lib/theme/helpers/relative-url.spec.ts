import { mocked } from 'ts-jest/utils';
import { MarkdownPlugin } from '../../plugin';
import { relativeUrl } from './relative-url';

jest.mock('../../plugin');
jest.mock('../theme');

describe(`relativeUrl helper`, () => {
  test(`should compile`, () => {
    mocked(MarkdownPlugin).location = 'interfaces/interface.md';
    const result = relativeUrl('modules/classes.md');
    expect(result).toMatchSnapshot();
  });
});
