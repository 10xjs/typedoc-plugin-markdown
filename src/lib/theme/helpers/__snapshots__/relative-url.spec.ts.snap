// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relativeUrl helper should compile 1`] = `"../modules/classes.md"`;
